File: 011_020/ContainerWithMostWater.py
# 011. Container With Most Water

class Solution:
    def maxArea(self, height):
        """
        :type height: List[int]
        :rtype: int
        """
        lo = 0
        hi = len(height) - 1
        maxArea = 0
        while lo < hi:
            area = (hi - lo) * min(height[lo], height[hi])
            maxArea = max(maxArea, area)
            if height[lo] < height[hi]:
                lo += 1
            else:
                hi -= 1

        return maxArea

File: go_solution/011_020/containerWithMostWater/containerWithMostWater.go
package containerWithMostWater
// 11. Container With Most Water
func maxArea(height []int) int {
	// 从头和尾部开始向内搜索
	// 因为i:=0,j;=len(height)-1时，长度是最长的，所以向内搜索的时候只用考虑高度更高的情况
	i, j := 0, len(height)-1
	rlt := 0
	for i < j {
		rlt = max(rlt, (j-i)*min(height[i], height[j]))
		if height[i] < height[j] {
			i++
		}else {
			j--
		}
	}
	return rlt
}

func max(a, b int) int {
	if a > b {
		return a
	}
	return b
}

func min(a, b int) int {
	if a > b {
		return b
	}
	return a
}